[optim] Add general documentation on our algorithm defaults

janeyx99 opened the pull request:
I added a section + table under Algorithms
https://docs-preview.pytorch.org/95391/optim.html?highlight=optim#module-torch.optim
<img width="725" alt="image" src="https://user-images.githubusercontent.com/31798555/221246256-99325a27-9016-407b-a9fe-404d61e41a82.png">


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95391

janeyx99 committed:
[optim] Add general documentation on our algorithm defaults

[ghstack-poisoned]

janeyx99 committed:
Update on "[optim] Add general documentation on our algorithm defaults"




[ghstack-poisoned]

janeyx99 committed:
Update on "[optim] Add general documentation on our algorithm defaults"




[ghstack-poisoned]

albanD commented on docs/source/optim.rst:
```diff
@@ -129,6 +129,50 @@ Algorithms
     Rprop
     SGD
 
+Many of our algorithms have various implementations optimized for performance, so
```
nit: they're not all optimized for performance. Some are performance, some readability/generality?

albanD commented on docs/source/optim.rst:
```diff
@@ -129,6 +129,50 @@ Algorithms
     SGD
 
+Many of our algorithms have various implementations optimized for performance, so
+we attempt to default to the faster foreach implementation if no particular
```
nit: you didn't really define what "foreach" mean at this point. This can just say we default to the generally fastest implementation for the current device.

albanD reviewed (commented):
Stylistic comments only.
Will need to make sure that the page renders well on the rendered doc.

janeyx99 committed:
Update on "[optim] Add general documentation on our algorithm defaults"

I added a section + table under Algorithms
https://docs-preview.pytorch.org/95391/optim.html?highlight=optim#module-torch.optim




[ghstack-poisoned]

janeyx99 committed:
Update on "[optim] Add general documentation on our algorithm defaults"

I added a section + table under Algorithms
https://docs-preview.pytorch.org/95391/optim.html?highlight=optim#module-torch.optim




[ghstack-poisoned]

albanD commented on docs/source/optim.rst:
```diff
@@ -129,6 +129,49 @@ Algorithms
+of computation all at once, thereby saving many sequential kernel calls. A few of our
+optimizers have even faster fused implementations, which fuse the big chunks of
+computation into one kernel. We can think of foreach implementations as fusing
+horizontally and fused implementations as fusing vertically.
```
as fusing vertically on top of that.

As fused is also doing the horizontal fusion that foreach is doing

albanD reviewed (approved):
One small nit but LGTM otherwise!

janeyx99 committed:
Update on "[optim] Add general documentation on our algorithm defaults"

I added a section + table under Algorithms
https://docs-preview.pytorch.org/95391/optim.html?highlight=optim#module-torch.optim
<img width="725" alt="image" src="https://user-images.githubusercontent.com/31798555/221246256-99325a27-9016-407b-a9fe-404d61e41a82.png">





[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot merge -f "mac tests are queuing for a long time, merging without waiting"
